pkg/gcp/machine_controller_util.go

```diff
@@ -379,15 +379,20 @@ func prepareErrorf(err error, format string, args ...interface{}) error {
 
 // ExtractProject returns the name of the project which is extracted from the secret
 func ExtractProject(credentialsData map[string][]byte) (string, error) {
-	serviceAccountJSON := extractCredentialsFromData(credentialsData, api.GCPServiceAccountJSON, api.GCPAlternativeServiceAccountJSON)
+	credentialsConfigJSON := extractCredentialsFromData(credentialsData, api.GCPServiceAccountJSON, api.GCPAlternativeServiceAccountJSON, api.GCPCredentialsConfig)
 
 	var j struct {
 		Project string `json:"project_id"`
 	}
-	if err := json.Unmarshal([]byte(serviceAccountJSON), &j); err != nil {
+	if err := json.Unmarshal([]byte(credentialsConfigJSON), &j); err != nil {
 		return "Error", err
 	}
-	return j.Project, nil
+
+	if j.Project != "" {
+		return j.Project, nil
+	}
+
+	return string(credentialsData["projectID"]), nil
```

Comment:
Is this map access guaranteed to always be successful?

Reply:
The previous implementation was relying that "project_id" is always in the secret data. If not it was returning an empty string and a `nil` error. That behaviour is persisted with this change, but maybe an error can be returned

Reply:
Since this code doesn't change the existing flow, I don't think it's a big issue

Reply:
I believe you need a project explicitly set in the client calls that you make in the GCP client.